src/System.Linq/tests/SelectManyTests.cs

```diff
@@ -477,5 +477,19 @@ public void ThrowOverflowExceptionOnConstituentLargeCounts(int[] counts)
             IEnumerable<int> iterator = counts.SelectMany(c => Enumerable.Range(1, c));
             Assert.Throws<OverflowException>(() => iterator.Count());
         }
+
+        [Fact]
+        public void CollectionInterleavedWithLazyEnumerables_ToArray_Regression()
+        {
+            var results = new[] { 4, 5, 6 }
+                .SelectMany(i => i == 5 ?
```

Comment:
Seems like we should have tests with other operators and with varying sizes, not just SelectMany/Concat and not just with one or a few items.

Reply:
Also, since various operators specialize on the source collection type, seems like we need tests on various source types.

Reply:
> not just SelectMany/Concat

I explained in my last PR's description that since Append/Prepend never use the incorrect returned `CopyPosition`, they shouldn't be affected by this bug.

Reply:
> I explained in my last PR's description that since Append/Prepend never use the incorrect returned CopyPosition, they shouldn't be affected by this bug.

Even if that's true of the current implementation, you're assuming the current implementation will never change.  Tests are important both to verify existing behavior and to help prevent future regressions.

Reply:
The bug is affecting only SelectMany() and Concat(). I've added tests for small and large data sets.